src/main.rs

```diff
@@ -4,70 +4,58 @@ use utf8parse::Receiver;
 use std::env;
 use std::fs::File;
 use std::io::Read;
+use std::str;
 use std::time::{Duration, Instant};
 
-struct BufferReceiver {
-    data: Vec<char>,
-}
-
-impl BufferReceiver {
-    pub fn new() -> Self {
-        BufferReceiver { data: vec![] }
-    }
-    pub fn done(self) -> Vec<char> {
-        self.data
-    }
-}
+struct Utf8Receiver;
 
-impl Receiver for BufferReceiver {
-    fn codepoint(&mut self, c: char) {
-        self.data.push(c);
+impl Receiver for Utf8Receiver {
+    fn codepoint(&mut self, _c: char) {
     }
     fn invalid_sequence(&mut self) {
         panic!("This is a performance benchmark. Use only valid UTF-8");
     }
 }
 
 trait BenchableUtf8Parser {
-    fn parse(bytes: &[u8]) -> Vec<char>;
+    fn parse(bytes: Vec<u8>) -> String;
 }
 
 struct BenchableTableParser;
 
 impl BenchableUtf8Parser for BenchableTableParser {
-    fn parse(bytes: &[u8]) -> Vec<char> {
+    fn parse(bytes: Vec<u8>) -> String {
         let mut p = utf8parse::Parser::new();
-        let mut r = BufferReceiver::new();
+        let mut r = Utf8Receiver;
         for b in bytes.iter() {
             p.advance(&mut r, *b);
         }
-        r.done()
+
+        // if we didn't panic, then utf8 is valid
+        unsafe {
+            String::from_utf8_unchecked(bytes)
+        }
     }
 }
 
 struct BenchableStdlibParser;
 
 impl BenchableUtf8Parser for BenchableStdlibParser {
-    fn parse(bytes: &[u8]) -> Vec<char> {
-        let mut ret = Vec::new();
-        let chars_it = unsafe {
-            std::str::from_utf8_unchecked(bytes).chars()
-        };
-        for c in chars_it {
-            ret.push(c);
-        }
-        ret
+    fn parse(bytes: Vec<u8>) -> String {
+        String::from_utf8(bytes).unwrap()
     }
```

Comment:
Actually I think this new code does not do any parsing: The internal String representation of Rust *is* utf-8. So all this does is check that the UTF-8 is valid and then declare that the respective memory shall onward be called a String. The parsing in the original code only happens because we *iterate* over the chars: Chars are *decoded* utf-8.